diff --git a/CMakeLists.txt b/CMakeLists.txt
index 93f044a17e..409dd86785 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,6 +19,16 @@ endif()
 
 project(CATBOOST LANGUAGES C CXX ASM)
 
+# Use OpenSSL from system instead of the version provided by conan.
+#  1. conan supplied FindOpenSSL.cmake provides openssl::openssl
+#  2. cmake supplied FindOpenSSL.cmake provides OpenSSL::SSL and OpenSSL::Crypto.
+# See https://github.com/conan-io/conan-center-index/issues/952
+# This patch is needed to reconcile the difference.
+find_package(OpenSSL REQUIRED)
+add_library(LinkOpenSSL INTERFACE)
+target_link_libraries(LinkOpenSSL INTERFACE OpenSSL::SSL OpenSSL::Crypto)
+add_library(openssl::openssl ALIAS LinkOpenSSL)
+
 set(BUILD_SHARED_LIBS Off)
 set(CMAKE_CXX_STANDARD 20)
 set(CMAKE_CXX_EXTENSIONS Off)
diff --git a/build/toolchains/clang.toolchain b/build/toolchains/clang.toolchain
index 7cb81ff061..a2fb977b04 100644
--- a/build/toolchains/clang.toolchain
+++ b/build/toolchains/clang.toolchain
@@ -1,11 +1,8 @@
-set(CMAKE_C_COMPILER clang)
-set(CMAKE_CXX_COMPILER clang++)
 
 set(CMAKE_C_STANDARD_LIBRARIES "-lc -lm")
 set(CMAKE_CXX_STANDARD_LIBRARIES "-lc -lm")
 
 set(CMAKE_CUDA_COMPILER nvcc)
-set(CMAKE_CUDA_HOST_COMPILER clang-14)
 
 set(CMAKE_CUDA_FLAGS_INIT "--allow-unsupported-compiler")
 
@@ -14,9 +11,6 @@ if (NOT APPLE)
   set(CMAKE_SHARED_LINKER_FLAGS -fuse-ld=lld)
 endif()
 
-set(ENV{CC} clang)
-set(ENV{CXX} clang++)
-
 set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
 set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
 set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
diff --git a/catboost/pytest/smoke_tests/classification_tutorial_cpu.py b/catboost/pytest/smoke_tests/classification_tutorial_cpu.py
index 0d140f566c..3ff63efbf7 100644
--- a/catboost/pytest/smoke_tests/classification_tutorial_cpu.py
+++ b/catboost/pytest/smoke_tests/classification_tutorial_cpu.py
@@ -42,7 +42,6 @@ except:
 
 #!pip install --user --upgrade catboost
 #!pip install --user --upgrade ipywidgets
-#!pip install shap
 #!pip install sklearn
 #!pip install --upgrade numpy
 #!jupyter nbextension enable --py widgetsnbextension
@@ -643,24 +642,6 @@ shap_values = shap_values[:,:-1]
 print(shap_values.shape)
 
 
-import shap
-shap.initjs()
-shap.force_plot(expected_value, shap_values[3,:], X.iloc[3,:], show=False)
-
-
-import shap
-shap.initjs()
-shap.force_plot(expected_value, shap_values[91,:], X.iloc[91,:], show=False)
-
-
-shap.summary_plot(shap_values, X, show=False)
-
-
-X_small = X.iloc[0:200]
-shap_small = shap_values[:200]
-shap.force_plot(expected_value, shap_small, X_small, show=False)
-
-
 ### Feature evaluation
 
 
diff --git a/catboost/python-package/pyproject.toml b/catboost/python-package/pyproject.toml
index 588bf56b9d..2d46df10a4 100644
--- a/catboost/python-package/pyproject.toml
+++ b/catboost/python-package/pyproject.toml
@@ -2,9 +2,7 @@
 build-backend = "setuptools.build_meta"
 requires = [
     "setuptools (>=70.1.0, <83.0)",
-    "cmake (>=3.24, <4.0)",
     "jupyterlab (>=3.0.6, <3.6.0)",
-    "conan (>=2.4.1, <3.0)",
     "cython ~= 3.0.10",
     "numpy < 3.0"
 ]
diff --git a/catboost/python-package/setup.py b/catboost/python-package/setup.py
index b4c1890b24..9ac885047b 100644
--- a/catboost/python-package/setup.py
+++ b/catboost/python-package/setup.py
@@ -124,7 +124,7 @@ def setup_hnsw_submodule(argv, extensions):
 
 
 def get_setup_requires(argv):
-    setup_requires = ['cmake (>=3.24, <4.0)', 'conan (>=2.4.1, <3.0)', 'cython ~= 3.0.10', 'numpy < 3.0']
+    setup_requires = ['cython ~= 3.0.10', 'numpy < 3.0']
 
     if ('build_widget' in argv) or (not ('--no-widget' in argv)):
         setup_requires += ['jupyterlab (>=3.0.6, <3.6.0)']
diff --git a/contrib/libs/cxxsupp/CMakeLists.linux-aarch64.txt b/contrib/libs/cxxsupp/CMakeLists.linux-aarch64.txt
index a660fcaeda..871123ae20 100644
--- a/contrib/libs/cxxsupp/CMakeLists.linux-aarch64.txt
+++ b/contrib/libs/cxxsupp/CMakeLists.linux-aarch64.txt
@@ -12,7 +12,12 @@
 # Please refer to the build instructions in the repository for more information about manual
 # changes in this file.
 
-add_subdirectory(libcxx)
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    add_subdirectory(libcxxcuda11)
+else ()
+    add_subdirectory(libcxx)
+endif()
+
 add_subdirectory(libcxxabi-parts)
 add_subdirectory(libcxxrt)
 add_subdirectory(builtins)
@@ -22,6 +27,16 @@ add_library(contrib-libs-cxxsupp INTERFACE)
 
 target_link_libraries(contrib-libs-cxxsupp INTERFACE
   contrib-libs-linux-headers
-  libs-cxxsupp-libcxx
 )
 
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxxcuda11
+    )
+else ()
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxx
+    )
+endif()
+
+
diff --git a/contrib/libs/cxxsupp/CMakeLists.linux-ppc64le.txt b/contrib/libs/cxxsupp/CMakeLists.linux-ppc64le.txt
index a660fcaeda..871123ae20 100644
--- a/contrib/libs/cxxsupp/CMakeLists.linux-ppc64le.txt
+++ b/contrib/libs/cxxsupp/CMakeLists.linux-ppc64le.txt
@@ -12,7 +12,12 @@
 # Please refer to the build instructions in the repository for more information about manual
 # changes in this file.
 
-add_subdirectory(libcxx)
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    add_subdirectory(libcxxcuda11)
+else ()
+    add_subdirectory(libcxx)
+endif()
+
 add_subdirectory(libcxxabi-parts)
 add_subdirectory(libcxxrt)
 add_subdirectory(builtins)
@@ -22,6 +27,16 @@ add_library(contrib-libs-cxxsupp INTERFACE)
 
 target_link_libraries(contrib-libs-cxxsupp INTERFACE
   contrib-libs-linux-headers
-  libs-cxxsupp-libcxx
 )
 
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxxcuda11
+    )
+else ()
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxx
+    )
+endif()
+
+
diff --git a/contrib/libs/cxxsupp/CMakeLists.linux-x86_64.txt b/contrib/libs/cxxsupp/CMakeLists.linux-x86_64.txt
index a660fcaeda..871123ae20 100644
--- a/contrib/libs/cxxsupp/CMakeLists.linux-x86_64.txt
+++ b/contrib/libs/cxxsupp/CMakeLists.linux-x86_64.txt
@@ -12,7 +12,12 @@
 # Please refer to the build instructions in the repository for more information about manual
 # changes in this file.
 
-add_subdirectory(libcxx)
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    add_subdirectory(libcxxcuda11)
+else ()
+    add_subdirectory(libcxx)
+endif()
+
 add_subdirectory(libcxxabi-parts)
 add_subdirectory(libcxxrt)
 add_subdirectory(builtins)
@@ -22,6 +27,16 @@ add_library(contrib-libs-cxxsupp INTERFACE)
 
 target_link_libraries(contrib-libs-cxxsupp INTERFACE
   contrib-libs-linux-headers
-  libs-cxxsupp-libcxx
 )
 
+if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 16)
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxxcuda11
+    )
+else ()
+    target_link_libraries(contrib-libs-cxxsupp INTERFACE
+      libs-cxxsupp-libcxx
+    )
+endif()
+
+
diff --git a/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64-cuda.txt b/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64-cuda.txt
index 896ef42d3e..db7f92127e 100644
--- a/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64-cuda.txt
+++ b/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64-cuda.txt
@@ -24,15 +24,7 @@ target_link_libraries(libs-cxxsupp-builtins PUBLIC
 )
 
 target_sources(libs-cxxsupp-builtins PRIVATE
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/extendbfsf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncdfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncxfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncsfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/trunctfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/chkstk.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/fp_mode.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-abi-assert.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-abi.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvdi2.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvsi2.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvti2.c
@@ -185,8 +177,6 @@ target_sources(libs-cxxsupp-builtins PRIVATE
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umoddi3.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umodsi3.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umodti3.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-libc-opt-memcpy-memmove.S
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-libc-opt-memset-memchr.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/crtbegin.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/crtend.c
 )
diff --git a/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64.txt b/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64.txt
index 896ef42d3e..db7f92127e 100644
--- a/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64.txt
+++ b/contrib/libs/cxxsupp/builtins/CMakeLists.linux-aarch64.txt
@@ -24,15 +24,7 @@ target_link_libraries(libs-cxxsupp-builtins PUBLIC
 )
 
 target_sources(libs-cxxsupp-builtins PRIVATE
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/extendbfsf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncdfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncxfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/truncsfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/trunctfbf2.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/chkstk.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/fp_mode.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-abi-assert.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-abi.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvdi2.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvsi2.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/absvti2.c
@@ -185,8 +177,6 @@ target_sources(libs-cxxsupp-builtins PRIVATE
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umoddi3.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umodsi3.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/umodti3.c
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-libc-opt-memcpy-memmove.S
-  ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/aarch64/sme-libc-opt-memset-memchr.S
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/crtbegin.c
   ${PROJECT_SOURCE_DIR}/contrib/libs/cxxsupp/builtins/crtend.c
 )
diff --git a/contrib/libs/protoc/src/google/protobuf/compiler/java/context.h b/contrib/libs/protoc/src/google/protobuf/compiler/java/context.h
index 57b994a8e3..ff03d28608 100644
--- a/contrib/libs/protoc/src/google/protobuf/compiler/java/context.h
+++ b/contrib/libs/protoc/src/google/protobuf/compiler/java/context.h
@@ -35,6 +35,7 @@
 #include <vector>
 
 #include "y_absl/container/flat_hash_map.h"
+#include "google/protobuf/compiler/java/field.h"
 #include "google/protobuf/compiler/java/helpers.h"
 #include "google/protobuf/compiler/java/options.h"
 #include "google/protobuf/port.h"
@@ -59,8 +60,6 @@ namespace protobuf {
 namespace compiler {
 namespace java {
 
-struct FieldGeneratorInfo;
-struct OneofGeneratorInfo;
 // A context object holds the information that is shared among all code
 // generators.
 class Context {
diff --git a/contrib/restricted/abseil-cpp-tstring/y_absl/container/internal/raw_hash_set.h b/contrib/restricted/abseil-cpp-tstring/y_absl/container/internal/raw_hash_set.h
index bb750bc7f8..ac6e55e140 100644
--- a/contrib/restricted/abseil-cpp-tstring/y_absl/container/internal/raw_hash_set.h
+++ b/contrib/restricted/abseil-cpp-tstring/y_absl/container/internal/raw_hash_set.h
@@ -3026,13 +3026,13 @@ class raw_hash_set {
   //   s.insert({"abc", 42});
   std::pair<iterator, bool> insert(init_type&& value)
       Y_ABSL_ATTRIBUTE_LIFETIME_BOUND
-#if Y_ABSL_INTERNAL_CPLUSPLUS_LANG >= 202002L && (!defined(_MSC_VER) || defined(__clang__))
+#if Y_ABSL_INTERNAL_CPLUSPLUS_LANG >= 202002L && (defined(__clang__) && (__clang_major__ >= 16))
     requires(!IsLifetimeBoundAssignmentFrom<init_type>::value)
 #endif
   {
     return emplace(std::move(value));
   }
-#if Y_ABSL_INTERNAL_CPLUSPLUS_LANG >= 202002L && (!defined(_MSC_VER) || defined(__clang__))
+#if Y_ABSL_INTERNAL_CPLUSPLUS_LANG >= 202002L && (defined(__clang__) && (__clang_major__ >= 16))
   std::pair<iterator, bool> insert(
       init_type&& value Y_ABSL_INTERNAL_ATTRIBUTE_CAPTURED_BY(this))
       Y_ABSL_ATTRIBUTE_LIFETIME_BOUND
diff --git a/util/generic/ptr.h b/util/generic/ptr.h
index 04cfe50a42..8e524925db 100644
--- a/util/generic/ptr.h
+++ b/util/generic/ptr.h
@@ -381,7 +381,7 @@ private:
     T* T_;
 };
 
-template <typename T, typename... Args, class = std::enable_if_t<std::is_constructible_v<T, Args...>>>
+template <typename T, typename... Args>
 [[nodiscard]] THolder<T> MakeHolder(Args&&... args) {
     return THolder<T>(new T(std::forward<Args>(args)...));
 }
@@ -1029,17 +1029,17 @@ using TAtomicSharedPtr = TSharedPtr<T, TAtomicCounter, D>;
 template <class T, class D = TDelete>
 using TSimpleSharedPtr = TSharedPtr<T, TSimpleCounter, D>;
 
-template <typename T, typename C, typename... Args, class = std::enable_if_t<std::is_constructible_v<T, Args...>>>
+template <typename T, typename C, typename... Args>
 [[nodiscard]] TSharedPtr<T, C> MakeShared(Args&&... args) {
     return new T{std::forward<Args>(args)...};
 }
 
-template <typename T, typename... Args, class = std::enable_if_t<std::is_constructible_v<T, Args...>>>
+template <typename T, typename... Args>
 [[nodiscard]] inline TAtomicSharedPtr<T> MakeAtomicShared(Args&&... args) {
     return MakeShared<T, TAtomicCounter>(std::forward<Args>(args)...);
 }
 
-template <typename T, typename... Args, class = std::enable_if_t<std::is_constructible_v<T, Args...>>>
+template <typename T, typename... Args>
 [[nodiscard]] inline TSimpleSharedPtr<T> MakeSimpleShared(Args&&... args) {
     return MakeShared<T, TSimpleCounter>(std::forward<Args>(args)...);
 }
